subprojects/core/src/main/java/org/gradle/api/internal/AbstractTask.java

```diff
@@ -321,6 +324,17 @@ public void run() {
         });
     }
 
+    @Override
+    public void onlyIf(final Provider<Boolean> provider) {
+        taskMutator.mutate("Task.onlyIf(Provider)", new Runnable() {
+            @Override
+            public void run() {
+                addProviderTaskDependency(provider);
```

Comment:
❓ It seems the task dependencies of this provider would be retained even if `setOnlyIf(() -> true)` is called afterwards, right?

Reply:
Hm, yes, it is. Seems like we should remove dependency if `onlyIfSpec` doesn't contains `ProviderBackedSpec` with `TaskProvider` within?  Looking crappy